### PR TITLE
[air] Add train/tune benchmark

### PR DESCRIPTION
**Pull request by xwjiang2010**
Signed-off-by: Xiaowei Jiang <xwjiang2010@gmail.com>

## Why are these changes needed?
Making sure that tuning multiple trials in parallel is not significantly slower than training each individual trials.
Some overhead is expected.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by krfricke — COMMENTED** [review]
Looks great! Have you been able to run it?

Can we also add a single node case? 1x1 with CPU=1 vs. 1x1 with CPU=8

File: release/air_tests/air_benchmarks/workloads/tune_benchmark.py
Comment (krfricke) [inline release/air_tests/air_benchmarks/workloads/tune_benchmark.py:37]: I think we'll have to use CPU=4 here to utilize all workers
```suggestion
            "resources_per_worker": {"CPU": 4},
```

Comment (xwjiang2010) [reply release/air_tests/air_benchmarks/workloads/tune_benchmark.py:37]: wait, we are using m5.2xlarge (8CPU) instance and each trial is taking 1 machine. All together there are 8 machines.

File: release/air_tests/air_benchmarks/workloads/benchmark_util.py
Comment (richardliaw) [inline release/air_tests/air_benchmarks/workloads/benchmark_util.py:130]: maybe we can just use the default python timeit

https://docs.python.org/3/library/timeit.html#python-interface

File: python/ray/train/torch/train_loop_utils.py
Comment (richardliaw) [inline python/ray/train/torch/train_loop_utils.py:478]: @amogkam Unless I set this to local_rank, the benchmark script will error out.

If you print `get_gpu_ids()`, you get (2, 3) (len=2). However, you still need to set the device to local_rank or else you get a runtime error cuda invalid device ordinal.

I took a look at the blame which made this change here: https://github.com/ray-project/ray/commit/029517a037b1219423ab45af79db1e9296bc39c7

which is obviously different in logic but not quite sure what you were trying to do in the previous logic.

Comment (amogkam) [reply python/ray/train/torch/train_loop_utils.py:478]: Yeah this is a bug…will be fixed by https://github.com/ray-project/ray/pull/26493

using local_rank directly won’t work for fractional or multiple gpus per worker

**Review by krfricke — APPROVED** [review]
LGTM, shall we wait until #26493 is merged?

**Comment by krfricke** [comment]
Also, do we have a successful run somewhere?

**Comment by richardliaw** [comment]
Do not merge yet, this is based off of #26493

**Comment by krfricke** [comment]
Latest run: https://buildkite.com/ray-project/release-tests-pr/builds/10002#018216bf-4b81-4483-af95-e1800e2a4397

File: release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py
Comment (krfricke) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: Actually it does not consistently pass, but it seems to be closer. It could be due to constant overhead with the relatively short training time. I'll increase once more.

**Comment by krfricke** [comment]
Another run: https://buildkite.com/ray-project/release-tests-pr/builds/10008

File: release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py
Comment (krfricke) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: With more epochs it seems to be closer, indicating a constant setup overhead: https://buildkite.com/ray-project/release-tests-pr/builds/10008

Comment (ericl) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: Isn't this an odd benchmark case? Most users would presumably be using multiple GPUs per worker (as big as possible), so the packing strategy is irrelevant.

Comment (richardliaw) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: Performance on these Data parallel systems require processes to own 1 gpu each, even with multiple gpus per node.

Comment (richardliaw) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: It's from  the underlying allreduce algorithm

Comment (amogkam) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: Wait this seems like a pretty serious bug? If STRICT_PACK is feasible then PACK should always equal STRICT_PACK.

Btw we recently changed the default to SPREAD cc @matthewdeng. As @ericl pointed about before, all of our benchmarks should be using default configs, no?

Comment (matthewdeng) [reply release/air_tests/air_benchmarks/workloads/tune_torch_benchmark.py:47]: > cc @xwjiang2010 @amogkam @richardliaw @ericl STRICT_PACK makes a huge difference. Without it tune would consistently be ~60% worse than train (failing the threshold), most likely to workers being scheduled on different nodes. With it we are consistently within the allowed 20% overhead, even with different training job sizes (e.g. more epochs).
>
> Should we raise this with core (improve "PACK" placement strategy)? In any case we should document this.

Just to double check here, was this under the assumption that the default strategy was PACK and not SPREAD? (I am changing the default back to PACK and will re-run this test with default strategy)

**Pull request merged**
